CRM/Activityreport/Data.php

```diff
@@ -99,52 +261,60 @@ private static function getAPIDateFilter($startDate, $endDate) {
    */
   private static function splitMultiValues(array $data, $totalOffset, $multiValuesOffset) {
     $result = array();
+    $date = NULL;
     $i = 0;
 
     foreach ($data as $key => $row) {
-      $multiValuesRows = array();
+      $activityDate = substr($row['Activity Date Time'], 0, 10);
+
+      if (!$date) {
+        $date = $activityDate;
+      }
+
+      if ($date !== $activityDate) {
+        $totalOffset--;
+        break;
+      }
+
+      $multiValuesRows = null;
       if (!empty(self::$multiValues[$key])) {
         $multiValuesFields = array_combine(self::$multiValues[$key], array_fill(0, count(self::$multiValues[$key]), 0));
 
-        $multiValuesRows = self::populateMultiValuesRow($row, $multiValuesFields, $multiValuesOffset, self::ROWS_TO_RETURN - $i);
+        $multiValuesRows = self::populateMultiValuesRow($row, $multiValuesFields, $multiValuesOffset, self::ROWS_CACHE_LIMIT - $i);
 
         $result = array_merge($result, $multiValuesRows['data']);
         $multiValuesOffset = 0;
       } else {
-        $result[] = $row;
+        $result[] = array_values($row);
       }
       $i = count($result);
 
-      if ($i === self::ROWS_TO_RETURN) {
+      if ($i === self::ROWS_CACHE_LIMIT) {
         break;
       }
-      $totalOffset++;
-    }
 
-    $header = self::getHeader();
-    $output = array(array_keys($header));
-    foreach ($result as $row) {
-      $output[] = array_values($row);
+      unset(self::$multiValues[$key]);
+
+      $totalOffset++;
     }
 
     return array(
-      array(
-        'info' => array(
-          'nextOffset' => !empty($multiValuesRows['info']['multiValuesOffset']) ? $totalOffset : $totalOffset + 1,
-          'multiValuesOffset' => !empty($multiValuesRows['info']['multiValuesOffset']) ? $multiValuesRows['info']['multiValuesOffset'] : 0,
-          'multiValuesTotal' => !empty($multiValuesRows['info']['multiValuesTotal']) ? $multiValuesRows['info']['multiValuesTotal'] : 0,
-        ),
-        'data' => $output,
+      'info' => array(
+        'date' => $date,
+        'nextOffset' => !empty($multiValuesRows['info']['multiValuesOffset']) ? $totalOffset : $totalOffset + 1,
+        'multiValuesOffset' => !empty($multiValuesRows['info']['multiValuesOffset']) ? $multiValuesRows['info']['multiValuesOffset'] : 0,
+        'multiValuesTotal' => !empty($multiValuesRows['info']['multiValuesTotal']) ? $multiValuesRows['info']['multiValuesTotal'] : 0,
       ),
+      'data' => $result,
```

Comment:
Whenever you return an array with a lot of information, it's a good sign that you should actually return an object instead. Try to see if you encapsulate some of the logic used to create this data into a class and then make this method return an instance of this class